src/server/pkg/metrics/metrics.go

```diff
@@ -12,53 +13,131 @@ import (
+
+type readBatchOfActions struct {
+	channel chan countableUserActions
+}
+
+var readBatchOfActionsChannel = make(chan *readBatchOfActions, 0)
+
+//IncrementUserAction updates a counter per user per action for an API method by name
+func IncrementUserAction(userID string, action string) {
+	incrementActionChannel <- &incrementUserAction{
+		action: action,
+		user:   userID,
+	}
+}
+
+func (r *Reporter) dbMetrics(metrics *Metrics) {
 	cursor, err := gorethink.Object(
 		"Repos",
-		gorethink.DB(pfsDbName).Table("Repos").Count(),
+		gorethink.DB(r.pfsDbName).Table("Repos").Count(),
 		"Commits",
-		gorethink.DB(pfsDbName).Table("Commits").Count(),
+		gorethink.DB(r.pfsDbName).Table("Commits").Count(),
 		"ArchivedCommits",
-		gorethink.DB(pfsDbName).Table("Commits").Filter(
+		gorethink.DB(r.pfsDbName).Table("Commits").Filter(
 			map[string]interface{}{
 				"Archived": true,
 			},
 		).Count(),
 		"CancelledCommits",
-		gorethink.DB(pfsDbName).Table("Commits").Filter(
+		gorethink.DB(r.pfsDbName).Table("Commits").Filter(
 			map[string]interface{}{
 				"Cancelled": true,
 			},
 		).Count(),
 		"Files",
-		gorethink.DB(pfsDbName).Table("Diffs").Group("Path").Ungroup().Count(),
+		gorethink.DB(r.pfsDbName).Table("Diffs").Group("Path").Ungroup().Count(),
 		"Jobs",
-		gorethink.DB(ppsDbName).Table("JobInfos").Count(),
+		gorethink.DB(r.ppsDbName).Table("JobInfos").Count(),
 		"Pipelines",
-		gorethink.DB(ppsDbName).Table("PipelineInfos").Count(),
-	).Run(dbClient)
+		gorethink.DB(r.ppsDbName).Table("PipelineInfos").Count(),
+	).Run(r.dbClient)
 	if err != nil {
 		protolion.Errorf("Error Fetching Metrics:%+v", err)
 	}
 	cursor.One(&metrics)
 }
 
-// ReportMetrics blocks and reports metrics, if modified, to the
-// given kubernetes client every 15 seconds.
-func ReportMetrics(clusterID string, kubeClient *kube.Client, address string, pfsDbName string, ppsDbName string) {
-	dbClient, err := db.DbConnect(address)
-	if err != nil {
-		protolion.Errorf("Error connected to DB when reporting metrics: %v\n", err)
-		return
-	}
+// ReportMetrics blocks and reports metrics every 15 seconds
+func (r *Reporter) ReportMetrics() {
+	reportingTicker := time.NewTicker(time.Second * 15)
 	for {
-		metrics := &Metrics{}
-		dbMetrics(dbClient, pfsDbName, ppsDbName, metrics)
-		externalMetrics(kubeClient, metrics)
-		metrics.ID = clusterID
-		metrics.PodID = uuid.NewWithoutDashes()
-		metrics.Version = version.PrettyPrintVersion(version.Version)
-		reportSegment(metrics)
-		<-time.After(15 * time.Second)
+		select {
+		case incrementAction := <-incrementActionChannel:
+			if userActions[incrementAction.user] == nil {
+				userActions[incrementAction.user] = make(countableActions)
+			}
+			val, ok := userActions[incrementAction.user][incrementAction.action]
+			if ok {
+				userActions[incrementAction.user][incrementAction.action] = val.(uint64) + uint64(1)
+			} else {
+				userActions[incrementAction.user][incrementAction.action] = uint64(0)
+			}
```

Comment:
You don't need to branch here, you can do this in 1 line with:

```
userActions[incrementAction.user][incrementAction.action] = userActions[incrementAction.user][incrementAction.action] + 1
```

Also this code has a bug in it, the first incrementAction will leave the value at `0` rather than `1`, so the counter will always be 1 lower than it should be.